### PR TITLE
Data sources: Refactor logic for naming new data sources

### PR DESCRIPTION
**Pull request by IevaVasiljeva**
**What is this feature?**

Move the logic for naming new data sources from frontend to backend. Reasons for this:
* it is not guaranteed that the user will get a full list of data sources in the frontend (due to lacking permissions, for instance), so we can't reliably find a unique name in the frontend;
* it also feels like more of a backend responsibility.

**Why do we need this feature?**

To allow all users with data source creator permissions to create data sources through the UI (even if they are not allowed to list any of the existing data sources).

**Who is this feature for?**

Any RBAC users.

**Which issue(s) does this PR fix?**:

Fixes https://github.com/grafana/identity-access-team/issues/474

**Special notes for your reviewer:**

There are some changes in the names added to new data sources:
1. Names are now based on plugin `type` and not plugin `name`. From testing, they seem pretty similar, but it is slightly less neat. For instance, it'll now be `prometheus` not `Prometheus`, and `grafana-timestream-datasource` not `V2-Timestream`. If you think that this is an issue, I could pass plugin name in DS creation request.
2. Numbering when a lot of data sources of the same type are created and left with the default names - it now goes up to 9 not 10 (just because the logic was slightly simpler for that). So it'll be `prometheus-9-9-3` not `Prometheus-10-10-1`. If this is an issue the code can be updated to match the previous logic.

Please check that:
- [ ] It works as expected from a user's perspective.
- [ ] If this is a pre-GA feature, it is behind a feature toggle.
- [ ] The docs are updated, and if this is a [notable improvement](https://grafana.com/docs/writers-toolkit/writing-guide/contribute-release-notes/#how-to-determine-if-content-belongs-in-a-whats-new-document), it's added to our [What's New](https://grafana.com/docs/writers-toolkit/writing-guide/contribute-release-notes/) doc.

### CODE REVIEW

**Review by Jguer — APPROVED** [review]
nice

File: pkg/services/datasources/service/datasource.go [hunk 1]
```diff
@@ -236,6 +249,28 @@ func (s *Service) AddDataSource(ctx context.Context, cmd *datasources.AddDataSou
...
+	currentDigit := 0
+
+	for dsNames[strings.ToLower(name)] {
+		if currentDigit%10 == 0 {
+			currentDigit = 1
+			name = name + "-x"
+		}
+		name = fmt.Sprintf("%s%d", name[:len(name)-1], currentDigit)
+		currentDigit++
+	}
```
Comment (IevaVasiljeva) [inline pkg/services/datasources/service/datasource.go:269]: Currently the code would produces names like `prometheus-9-9-9-3`. This is sticking to how the old logic worked. Alternatively, we could change it to just keep incrementing the number (eg, we could have `prometheus-1234`). It seems simpler to me, but I don't know what the original reasoning was.

**Review by leventebalogh — APPROVED** [review]
LGTM 👍

File: pkg/services/datasources/service/datasource.go (hunk 1 above)
Comment (andresmgot) [reply pkg/services/datasources/service/datasource.go:269]: I guess that it is/was compared in alphabetical order?

**Review by andresmgot — APPROVED** [review]
LGTM

File: pkg/services/datasources/service/datasource.go (hunk 1 above)
Comment (IevaVasiljeva) [reply pkg/services/datasources/service/datasource.go:269]: Good question! I just realised that the old numbering would not have been in alphabetical order (`prometheus-8` is alphabetically after `prometheus-10-1`, but would have been created before it). The new numbering (counting up to 9 before hyphenating again) would result in correct alphabetical order.

That said, the more I think about it, the better it seems to just keep incrementing the number without adding more hyphens (so `prometheus-9`, `prometheus-10`, `prometheus-11` etc). I'll rewrite the code to do that. Let me know if you or @leventebalogh (the author of the original code) have any objections :)

**Pull request merged**